### PR TITLE
Support comments on ms sql server Fixes: #7844 (in progress, do not merge)

### LINKED ISSUE #7844: Support comments on SQL Server
### Describe the use case

It appears that support for comments was implemented in https://github.com/sqlalchemy/sqlalchemy/commit/fadb8d61babb76ef7bdbc98279096a8900c7328d, however it wasn't implemented for SQL Server - https://github.com/sqlalchemy/sqlalchemy/issues/1546#issuecomment-441906255

```python
 In[14]: engine.dialect
Out[14]: <sqlalchemy.dialects.mssql.pyodbc.MSDialect_pyodbc at 0x1db4d5289c8>

 In[15]: engine.dialect.supports_comments
Out[15]: False
```

It would be great if comments were also supported for SQL Server.

### Databases / Backends / Drivers targeted

MSSQL

### Example Use

NA

### Additional context

Just opening an issue to document and track this feature request as it took me a while to understand why the comments weren't working on my Azure SQL Database.

### PR DESCRIPTION
**Pull request by DanielHall977**
### Description
Tweaks requirements file to test mssql comment reflection.
Successfully adds table comments 

Fixes: https://github.com/sqlalchemy/sqlalchemy/issues/7844
### Checklist
- [ ] Fix failing test in QuotedNameArgumentTest
- [ ] Does it support column comments? 
- [ ] Is there valid test code for column comments?

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X  ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/dialects/mssql/base.py [hunk 1]
```diff
@@ -2607,6 +2607,72 @@ def visit_computed_column(self, generated):
...
+        schema_name = schema if schema else self.dialect.default_schema_name
+        return """
+           execute sp_addextendedproperty 'MS_Description', {0}, 'schema', {1}, 'table', {2};
+           """.format(
+            self.sql_compiler.render_literal_value(
+                create.element.comment, sqltypes.String()
+            ),
+            quoted_name(schema_name, quote=True),
+            quoted_name(
+                self.preparer.format_table(create.element, use_schema=False),
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mssql/base.py:2621]: here I think you can use return value of format table directly?

File: lib/sqlalchemy/dialects/mssql/base.py [hunk 2]
```diff
@@ -2607,6 +2607,72 @@ def visit_computed_column(self, generated):
             text += " PERSISTED"
         return text
 
+    def visit_set_table_comment(self, create):
+        schema = self.preparer.schema_for_object(create.element)
+        schema_name = schema if schema else self.dialect.default_schema_name
+        return """
+           execute sp_addextendedproperty 'MS_Description', {0}, 'schema', {1}, 'table', {2};
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mssql/base.py:2614]: @zzzeek a ddl cannot specify parameters rights? I believe this execute sql should support them, but we cannot pass parameters using a ddl at the moment, right?

**Review by CaselIT — CHANGES_REQUESTED** [review]
That's a good start, I've left some suggestions

thanks for working on it!

File: lib/sqlalchemy/dialects/mssql/base.py (hunk 2 above)
Comment (DanielHall977) [reply lib/sqlalchemy/dialects/mssql/base.py:2614]: In the other dialects instances of visit_set_table_comment use string formatting.

File: lib/sqlalchemy/dialects/mssql/base.py (hunk 1 above)
Comment (DanielHall977) [reply lib/sqlalchemy/dialects/mssql/base.py]: done, it was something I added to try to get past the QuotedNameArgumentTest unit tests

File: lib/sqlalchemy/dialects/mssql/base.py [hunk 3]
```diff
@@ -2607,6 +2607,72 @@ def visit_computed_column(self, generated):
...
+           """.format(
+            self.sql_compiler.render_literal_value(
+                create.element.comment, sqltypes.String()
+            ),
+            quoted_name(schema_name, quote=True),
```
[the remark this answers is not in the document]
Comment (DanielHall977) [reply lib/sqlalchemy/dialects/mssql/base.py]: done

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by CaselIT** [comment]
@zzzeek not sure why it's the gerrit is not working. do you have clues?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by zzzeek** [comment]
i can't really tell what the problem is

```
[2022-07-30T17:10:03.914Z] [ReceiveCommits-1[java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@723f8318[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@3cbfa31b[Wrapped task = com.google.gerrit.server.logging.LoggingContextAwareRunnable@1bcddd62]]]-for-HTTP POST /sqlalchemy/sqlalchemy/git-receive-pack (sqla-tester from 45.79.163.59)] WARN  com.google.gerrit.server.git.receive.ReceiveCommits : Skipping change updates on sqlalchemy/sqlalchemy because ref update failed: REJECTED_OTHER_REASON commit 82e9e9d: missing Change-Id in message footer [CONTEXT RECEIVE_ID="sqlalchemy/sqlalchemy-1659201003890-8e09caf9" project="sqlalchemy/sqlalchemy" ]

```

maybe somethign in the commit message is interfering

**Comment by zzzeek** [comment]
can someone force-push a squash of these changes to the PR here and make sure the commit message is absolutely clean of any weird formatting symbols ?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 540f4eb6395f9feed4b4240e3d22f539021948e9: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 540f4eb6395f9feed4b4240e3d22f539021948e9 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 540f4eb6395f9feed4b4240e3d22f539021948e9 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

this should be all done.

the reflection of the comments on the columns seems to not have noticeable slowdowns, at least when tested running

  python ./test/perf/many_table_reflection.py --db mssql+pyodbc://scott:tiger^5HHH@127.0.0.1:1433/test?driver=ODBC+Driver+17+for+SQL+Server  --table-number 100 --test columns

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

some leftover things, will fix

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

* lib/sqlalchemy/dialects/mssql/base.py (line 3449): leftover
* lib/sqlalchemy/dialects/mssql/information_schema.py (line 10): relative import
* lib/sqlalchemy/dialects/mssql/information_schema.py (line 240): update documentation

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

adding gord

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Comment by sqla-tester** [comment]
**Gord Thompson** (gordthompson) wrote:

Current code fails if the comment contains Unicode characters that cannot be mapped to the SBCS character set for the database because the `execute sp_addextendedproperty` statement passes the comment as 'string' instead of as N'string':

import sqlalchemy as sa

engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql_199")

tbl = sa.Table(
    "change_4016",
    sa.MetaData(),
    sa.Column(
        "id", sa.Integer, primary_key=True, autoincrement=False, comment="ΟΠΑ!"
    ),
)
tbl.drop(engine, checkfirst=True)
tbl.create(engine)
"""SQL received by SQL Profiler
CREATE TABLE change_4016 (
    id INTEGER NOT NULL, 
    PRIMARY KEY (id)
)
execute sp_addextendedproperty 'MS_Description', 'ΟΠΑ!', 'schema', [dbo], 'table', change_4016, 'column', id
"""

insp = sa.inspect(engine)
cols = insp.get_columns("change_4016")
print(repr(cols[0]["comment"]))  # '???!'


View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

yes, I noticed.

do you have suggestion on how to pass it as nvarchar?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4016 has been **merged**. Congratulations! :)

**Comment by gjolmej** [comment]
Is this currently in any available version of SQL alchemy? I can't find it.

**Comment by CaselIT** [comment]
It's in the sqlalchemy 2 betas that are on pypi

**Pull request closed without merging**